pkg/cache/multi_namespace_cache.go

```diff
@@ -174,9 +189,28 @@ func (c *multiNamespaceCache) List(ctx context.Context, list client.ObjectList,
 		// The last list call should have the most correct resource version.
 		resourceVersion = accessor.GetResourceVersion()
 	}
+
+	uniqueItems := removeDuplicates(allItems)
 	listAccessor.SetResourceVersion(resourceVersion)
 
-	return apimeta.SetList(list, allItems)
+	return apimeta.SetList(list, uniqueItems)
+}
+
+// removeDuplicates removes the duplicate objects obtained from all namespaces so that
+// the resulting list has objects with unique name and namespace.
+func removeDuplicates(items []runtime.Object) []runtime.Object {
```

Comment:
Here, I am filtering objects based on namespace and name. Since there is no client passed, we can't find if the object is specifically cluster scoped. An object of the same name can be created in multiple namespaces, but I suppose the combination of name+namespace is always unique. If there is a better way to do it, please let me know.

Reply:
Sorry, I am a bit confused by this. So today we get duplicate resources when we have a cache for multiple namespaces. Doesn't that imply that we already have a cluster-scoped cache attached to each of the namespaced caches? If so, why do we create another, explicit one?

IMHO we should just pass on the restmapper from the cache constructor, figure out if an object is cluster scoped and handle that case explicitly.

Reply:
To me, having a `removeDuplicates` looks like we did something wrong, because it implies that we end up creating redundant caches

Reply:
You're suggesting something like the following then?
```go
type multiNamespaceCache struct {
	namespaceToCache map[string]Cache
	Scheme           *runtime.Scheme
	RESTMapper       meta.RESTMapper
}

func (c *multiNamespaceCache) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
    gvk := list.ObjectKind().GroupVersionKind()
	mapping, _ := c.RESTMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
	if mapping.Scope.Name() == meta.RESTScopeNameRoot {
		// Handle cluster-scoped case
		return list
	}

	// Namespace-scoped operations
	...
}
```

Reply:
Yeah, exactly

Reply:
Great. The only other suggestion I have is to maintain GVK-to-scope mapping as `map[schema.GroupVersionKind]RESTScopeName`, since `RESTMapper.RESTMapping()` is potentially an expensive operation to perform every time `List()` is called.

Reply:
The restmapper will already only do an api call if gets a NoKindMatchError (i.E. it doesn't have the mapping), so I don't think we will need to cache this: https://github.com/kubernetes-sigs/controller-runtime/blob/e388e1e1f5a745166955442294bb8aede1f761e2/pkg/client/apiutil/dynamicrestmapper.go#L252

Reply:
True, but the `RESTMapping()` method for the static RESTMapper is still [much more expensive](https://github.com/kubernetes/apimachinery/blob/v0.20.4/pkg/api/meta/restmapper.go#L451) than a map lookup in the worst case. I’m almost splitting hairs on efficiency here, but not quite.